antipackage.py

```diff
@@ -1,130 +1,401 @@
-        top, username, repo, modname = self._parse_fullname(fullname)
-        return os.path.join(self.base_dir, top, username, repo, modname+'.py')
-
-    def _make_package(self, fullname):
-        top, username, repo, modname = self._parse_fullname(fullname)
-        if repo is not None:
-            repo_path = os.path.join(self.base_dir, top, username, repo)
-            self._setup_package(repo_path)
-        if username is not None:
-            user_path = os.path.join(self.base_dir, top, username)
-            self._setup_package(user_path)
-        if top is not None:
-            top_path = os.path.join(self.base_dir, top)
-            self._setup_package(top_path)
-        if modname is not None:
+    def waiting(self):
+        """Ready for import?"""
+        if not self.repo or len(self.comps)>3:
+            return True
+        return False
+
+    def exists(self):
+        """Data and self.path exist?"""
+        self.data = data('/'.join(self.index[1:]))
+        return os.path.exists(self.path) and self.data
+
+    def update(self):
+        """Use existing files?"""
+        pathlist = self.index[1:]
+        reponame = '.'.join(pathlist)
+        for key in self.data:
+            if key in ('branch','tag'):
+                repo = GitHubRepo(pathlist, key, self.data[key])
+                if repo.sha != self.data['commit']['sha']:
+                    print('Updating repo: %s' % reponame)
+                    self.data['commit']['sha'] = repo.sha
+                    return True
+        print('Using existing version of: %s' % reponame)
+        return False
+
+    def install(self):
+        """Download and install new repository at self.path"""
+        url = self.zip()
+        print("Downloading from: %s" % url)
+        sys.stdout.flush()
+        temp_file, response = urlretrieve(url, reporthook=display_download)
+        zf = zipfile.ZipFile(temp_file)
+        fname = zf.filename
+        with TemporaryDirectory() as td:
+            zf.extractall(td)
+            contents = os.listdir(td)
+            for name in contents:
+                if name.startswith(self.username):
+                    tmp_repo = name
+            if os.path.exists(self.path):
+                shutil.rmtree(self.path)
+            os.rename(td+'/'+tmp_repo, self.path)
+        self._install_init()
+
+    def zip(self):
+        """Returns url to the reopsitory's .zip file"""
+        path = ['repos',self.username, self.repo,'zipball']
+        if self.data:
+            src = self.data['commit']['sha']
+        else:
+            pin('/'.join(self.index[1:]))
+            src = 'master'
+        path.append(src)
+        return 'https://api.github.com/'+'/'.join(path)
+
+class GitHubRepo(object):
+
+    host = {'top': 'https://api.github.com/repos/{1}/{2}/',
+            'branch': 'branches/{0}',
+            'sha': 'commits/{0}',
+            'tag': 'tags'}
+
+    def __init__(self, pathlist, name='branch', value='master'):
```

Comment:
I think the `__init__` method should not do the network requests. Let's separate out the init from the `_git_init` stuff. I also think we should refactor the signatures a bit:

```python
def __init__(self, username, repo):
    # no network requests done here

def fetch_metadata(self, sha=None, tag=None, branch=None):
    # This is where the network request will be done.
    # This will make it easier for the code calling this class to catch exceptions that happen
    # in the network requests.
```